oak_functions_service/src/lookup_htbl.rs

```diff
@@ -0,0 +1,470 @@
+// bust out of on-chip cache, so speed is limited by cache misses, not the actual operations such
+// as hashing and bit manipulation.
+//
+// The k/v pairs are loaded before any are read, and once they are loaded cannot be modified.  We
+// take advantage of these restrictions to build a hash table with far less memory overhead per k/v
+// pair (20 bytes vs ~128 bytes for hashbrown<Vec<u8>, Vec<u8>>), and similar speed k/v lookups
+// (~90ns on manual tests).  For small average k/v pairs, of around 64 bytes total, this hash table
+// is about 2.5X more memory efficient.
+
+use alloc::{vec, vec::Vec};
+use core::mem;
+
+use bytes::Bytes;
+
+// To save memory, we use 5 byte array index values instead of 8 bytes.  This saves 12 bytes per
+// k/v pair.  It is referred to as u40 below.
+const INDEX_SIZE: usize = 5;
+// Size of an entry in the hash table.
+const ENTRY_SIZE: usize = INDEX_SIZE + 1;
+
+// The hash table consists of 2 vectors.  The first is a table of "entries", which is a packed
+// array of entries with the following layout:
+//
+//    hash_byte: u8
+//    data_index: u40  // A 5-byte index, so we support up to 1TiB of data.
+//
+// This is a "swiss" table meaning that instead of the usual "buckets" of entries, the entries are
+// collapsed into the hash table itself.  By adding a 1-byte hash, we can eliminate over 99.5% of
+// key comparisons to the wrong keys.  This is important for speed because comparing keys causes a
+// page fault usually, while comparing thie 1-byte hash is usually an L1 cache hit.  The number of
+// cache misses per lookup should be close to 2 for values found in the table, and 1 for values not
+// in the table.
+//
+// The second table contains fixed k/v entries in the following format:
+//
+//     key_len: u40
+//     key: [u8]
+//     value_len: u40
+//     value: [u8]
+//
+// Another memory savings comes from using a swiss table with exactly a 60% load factor.  Swiss
+// tables require a significant portion of the table be empty, and normally as the table grows,
+// they dynamically double the size.  At least when using the alloc crate, this resize causes a new
+// table to be allocated at 2X the size, data is copied from the old table to the new, and the old
+// table is freed.  This results in a 3X expansion in memory usage vs the prior array size.  In
+// contrast, this scheme has only 40% of the table empty and is never resized.
+//
+// This reduce scheme currently limits the number of entries to 2^32.
+#[derive(Default)]
+pub struct LookupHtbl {
+    table: Vec<u8>,
+    data: Vec<u8>,
+    max_entries: usize,       // The number at which we must grow the table.
+    allocated_entries: usize, // This is self.table.len() * ENTRY_SIZE.
+    used_data: usize,
+    used_entries: usize,
+}
+
+enum LookupResult {
+    Found(usize),        // (data_index: u40)
+    NotFound(usize, u8), // (table_index: u40, hash_byte: u8)
+}
+
+impl LookupHtbl {
+    pub fn new() -> LookupHtbl {
```

Comment:
Rust style nit: this is just `Default::default()`, no need to implement `new()` in this case. Just use `default()`.

Reply:
Done.